### PR TITLE
Add handling for 1 month intro discounts in Jetpack 2 year upsell

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
## Proposed Changes

* Ignore 1 month intro discounts when calculating the 2 year upsell discount percentage

Slack: p1694166055541249-slack-C02LK1W8T4Z

## Testing Instructions

1. Use the Calypso live link or checkout this branch locally
2. Go to `/checkout/jetpack/jetpack_social_advanced_yearly`
3. Make sure the upsell discount percentage is calculated correctly
image
4. Check other products to make sure they still work as well
(stats)
image
(backup)
image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/checkout/src/components/item-variation-picker/util.ts [hunk 1]
```diff
@@ -14,6 +14,11 @@ export function getItemVariantCompareToPrice(
 		return undefined;
 	}
 
+	// Ignore a 1 month discount when calculating the discount percentage
+	if ( compareTo.introductoryInterval === 1 && compareTo.introductoryTerm === 'month' ) {
+		return compareTo.priceBeforeDiscounts * 2;
+	}
+
```
Comment (jboland88) [inline client/my-sites/checkout/src/components/item-variation-picker/util.ts]: Does the multiplication by 2 only apply if the `variant` has a 2 year term? it seems to have an interesting effect on showing discounts in the plan select drop-down. The monthly plan shows as a 94% discount, which is not correct.

Screenshot 2023-09-08 at 11 56 35 AM

**Review by jboland88 — CHANGES_REQUESTED** [review]
This does appear to fix the sidebar, but is creating a new issue in the plan term drop-down.

File: client/my-sites/checkout/src/components/item-variation-picker/util.ts (hunk 1 above)
Comment (CodeyGuyDylan) [reply client/my-sites/checkout/src/components/item-variation-picker/util.ts]: Oh nice catch, I didn't account for that thank you, I'll get that fixed

Comment (CodeyGuyDylan) [reply client/my-sites/checkout/src/components/item-variation-picker/util.ts]: I got a fix and everything seems to be working. As we've always talked about before though it does seem this along with pricing stuff in general could be refactored in a more robust way. But I think this works for now to solve the immediate issue

**Review by sirbrillig — COMMENTED** [review]
Is it possible to write a unit test for this behavior? We have [a suite for the variant picker here](https://github.com/Automattic/wp-calypso/blob/2770fae63ecdc3de8c7edbb102b904559450ac60/client/my-sites/checkout/src/test/composite-checkout-variant-picker-dropdown.tsx#L50).

**Comment by CodeyGuyDylan** [comment]
> Is it possible to write a unit test for this behavior? We have [a suite for the variant picker here](https://github.com/Automattic/wp-calypso/blob/2770fae63ecdc3de8c7edbb102b904559450ac60/client/my-sites/checkout/src/test/composite-checkout-variant-picker-dropdown.tsx#L50).

Good thinking, I will add one 😄

File: client/my-sites/checkout/src/components/item-variation-picker/util.ts [hunk 2]
```diff
@@ -14,6 +14,15 @@ export function getItemVariantCompareToPrice(
 		return undefined;
 	}
 
+	// Ignore a 1 month discount when calculating the discount percentage
+	if (
+		variant.termIntervalInMonths === 24 &&
```
Comment (jboland88) [inline client/my-sites/checkout/src/components/item-variation-picker/util.ts:19]: Small nitpick, this is the only place I see `termIntervalInMonths` used in the function. Not sure if it makes sense to make all the checks consistent where you check the interval and term? (I think this works fine, was just curious)

**Review by jboland88 — APPROVED** [review]
This LGTM - I think a unit test could be helpful per @sirbrillig 's comment

**Comment by CodeyGuyDylan** [comment]
> This LGTM - I think a unit test could be helpful per @sirbrillig 's comment

Working on a unit test now, I don't think the variant dropdown is being populated correctly when 2 year plans are involved during the tests so I am debugging that. I think I'll merge this one to fix the issue and add a unit test after 🤔

File: client/my-sites/checkout/src/components/item-variation-picker/util.ts (hunk 2 above)
Comment (CodeyGuyDylan) [reply client/my-sites/checkout/src/components/item-variation-picker/util.ts:19]: I will merge this for now but I will look at improvements in a separate PR where I am adding a test

**Pull request merged**